engine/OthelloEngine.py

```diff
@@ -73,11 +73,25 @@ def update_board(self, move):
       
    # Check for end condition
    def check_end(self):
-      # TODO Task 8 Below:
       # Check the board to see if the game can continue
       # If the game is over return the winner: 'W', 'B', or 'T'
       # Otherwise, return None
-      pass
+
+      # "The dark player moves first", https://en.wikipedia.org/wiki/Reversi#Rules
+      current_player = 'W' if self.turn_number % 2 else 'B'
+
+      if len(get_all_moves(self.game_state, current_player)) != 0:
+         return None
```

Comment:
There is one more case I believe: let's say it's black's turn, but black cannot move anywhere. That means that black must forfeit their turn, and then white can attempt to make a move. Thus, even if current_player can't make any moves, that doesn't necessarily mean the game is over, you need to check the other player as well.

Reply:
Nice catch! You're absolutely right: "When neither player can move, the game ends" from the same Wikipedia page. I wasn't aware of the turn forfeiture game mechanic.